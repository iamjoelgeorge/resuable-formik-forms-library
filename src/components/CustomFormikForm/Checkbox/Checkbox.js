/* eslint-disable */
import React, { useState } from 'react';

import { Field } from 'formik';
import PropTypes from 'prop-types';

import styles from './Checkbox.module.scss';
import { joinClassNames } from '../../../utils/utils';
import ErrorText from '../ErrorText/ErrorText';
import SlidingLabel from '../SlidingLabel/SlidingLabel';
import ToolTip from '../ToolTip/ToolTip';
import AdditionalInfo from '../AdditionalInfo/AdditionalInfo';

const Checkbox = (props) => {
  const {
    name,
    optionLabel,
    mainLabel,
    formik,
    isRequired,
    isDisabled,
    helpLink,
    helpLinkText,
    tooltipLink,
    tooltipLinkText,
    optionalText,
    mainLabelTooltipBoxHeading,
    mainLabelTooltipBoxDescription,
    mainLabelTooltipBoxDescriptionElement,
    optionLabelTooltipBoxHeading,
    optionLabelTooltipBoxDescription,
    optionLabelTooltipIconChildElement,
    ...rest
  } = props;
  const [isFocused, setIsFocused] = useState(false);

  const { errors } = formik;

  const userHasVisitedTheInputField = formik.touched[name];
  const inputFieldHasErrors = errors[name];
  const addErrorClassesToLabelAndInput = !!userHasVisitedTheInputField && !!inputFieldHasErrors;

  const labelTextClasses = isFocused
    ? joinClassNames([styles.labelSpanElement, styles.focused])
    : styles.labelSpanElement;

  const showOptionTooltipIcon =
    optionLabelTooltipBoxHeading ||
    optionLabelTooltipBoxDescription ||
    optionLabelTooltipIconChildElement;

  const handleFocus = () => {
    setIsFocused(true);
  };

  const handleBlur = () => {
    setIsFocused(false);
  };

  return (
    <div className={styles.container}>
      <SlidingLabel
        customClass={styles.componentHeading}
        label={mainLabel}
        inputEntered
        htmlFor={name}
        showErrorStyle={addErrorClassesToLabelAndInput}
        tooltipIconBoxHeading={mainLabelTooltipBoxHeading}
        tooltipIconBoxDescription={mainLabelTooltipBoxDescription}
        tooltipIconChildElement={mainLabelTooltipBoxDescriptionElement}
        inputIsRequired={isRequired}
      />
      <div className={styles.optionWithTooltip}>
        <label className={styles.optionLabel}>
          <Field
            data-testid={`checkbox-${name}`}
            className={styles.checkboxInput}
            type='checkbox'
            name={name}
            {...rest}
            onFocus={handleFocus}
            onBlur={handleBlur}
            disabled={isDisabled}
          />

          <span className={styles.presentationalCheckbox}>
            <span className={styles.checkboxChecked} />
          </span>

          {optionLabel && <span className={labelTextClasses}>{optionLabel}</span>}
        </label>

        {showOptionTooltipIcon && (
          <ToolTip
            heading={optionLabelTooltipBoxHeading}
            description={optionLabelTooltipBoxDescription}
            descriptionElement={optionLabelTooltipIconChildElement}
            containerClass={styles.tooltip}
          />
        )}
      </div>

      {addErrorClassesToLabelAndInput && (
        <ErrorText fieldName={name} containerClass={styles.error} />
      )}

      <AdditionalInfo
        optionalText={optionalText}
        helpLinkText={helpLinkText}
        helpLink={helpLink}
        tooltipLinkText={tooltipLinkText}
        tooltipLink={tooltipLink}
        customClass={styles.optionalText}
      />
    </div>
  );
};

Checkbox.propTypes = {
  name: PropTypes.string.isRequired,
  optionLabel: PropTypes.string,
  mainLabel: PropTypes.string,
  isRequired: PropTypes.bool,
  isDisabled: PropTypes.bool,
  helpLink: PropTypes.string,
  helpLinkText: PropTypes.string,
  tooltipLink: PropTypes.string,
  tooltipLinkText: PropTypes.string,
  optionalText: PropTypes.string,
  mainLabelTooltipBoxHeading: PropTypes.string,
  mainLabelTooltipBoxDescription: PropTypes.string,
  mainLabelTooltipBoxDescriptionElement: PropTypes.element,
  optionLabelTooltipBoxHeading: PropTypes.string,
  optionLabelTooltipBoxDescription: PropTypes.string,
  optionLabelTooltipIconChildElement: PropTypes.element,
};

Checkbox.defaultProps = {
  optionLabel: '',
  mainLabel: '',
  isRequired: false,
  isDisabled: false,
  helpLink: '',
  helpLinkText: '',
  tooltipLink: '',
  tooltipLinkText: '',
  optionalText: '',
  mainLabelTooltipBoxHeading: '',
  mainLabelTooltipBoxDescription: '',
  optionLabelTooltipBoxHeading: '',
  optionLabelTooltipBoxDescription: '',
};

export default Checkbox;
