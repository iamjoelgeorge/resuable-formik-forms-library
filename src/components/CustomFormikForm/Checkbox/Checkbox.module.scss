@import '../../../theme/variables.scss';
@import '../../../theme/utils.scss';

* {
  box-sizing: border-box;
}

.container {
  position: relative;
  margin-top: 40px;

  .componentHeading {
    top: -20px;
    left: 0;
  }

  .optionLabel {
    position: relative;
    display: flex;
    justify-content: flex-start;
    align-items: center;
    font-size: em(14);
    line-height: 14px;
    width: fit-content;
    color: $primaryDark;
    cursor: pointer;
  }

  .tooltip {
    margin-left: 4px;
  }

  .optionWithTooltip{
    display: flex;
  }

  .checkboxInput {
    position: absolute;
    left: -999em;

    &:disabled + span.presentationalCheckbox,
    &:disabled + span.presentationalCheckbox + span.labelSpanElement {
      color: $secondaryGrey;
      cursor: not-allowed;
    }
  }

  .presentationalCheckbox {
    display: flex;
    justify-content: center;
    align-items: center;
    padding: em(1);
    margin-top: 2px;
    width: em(14);
    height: em(14);
    margin-right: em(10);
    border: 1px solid $lighterDarkGrey;
    border-radius: 3px;
    box-shadow: 0 2px 2px $borderGrey;
  }

  .checkboxChecked {
    display: inline-block;
    width: 100%;
    height: 100%;
    border-radius: 2px;
  }

  .labelSpanElement {
    margin-top: em(3);
  }

  .checkboxInput:checked + .presentationalCheckbox .checkboxChecked {
    background-color: $brightPurple;
  }

  .focused {
    border-color: $brightPurple;
    border-style: dashed;
  }

  .optionalText {
    padding-left: 0;
  }
}
