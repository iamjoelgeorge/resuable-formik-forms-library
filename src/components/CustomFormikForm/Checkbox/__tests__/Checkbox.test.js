import React from 'react';
import { fireEvent, wait, render, cleanup } from '@testing-library/react';
import "@testing-library/jest-dom/extend-expect";

import renderer from 'react-test-renderer';
import { useField } from 'formik';

import Checkbox from '../Checkbox';

jest.mock('formik');

const mockMeta = {
  touched: false,
  error: '',
  initialError: '',
  initialTouched: false,
  initialValue: '',
  value: '',
};

const mockField = {
  value: '',
  checked: false,
  onChange: jest.fn(),
  onBlur: jest.fn(),
  multiple: undefined,
  name: 'test',
};

useField.mockReturnValue([mockField, mockMeta]);

const defaultProps = {
  name: 'license',
  isRequired: false,
  isDisabled: false,
  optionLabel: 'Do you have a license?',
  mainLabel: 'License',
  helpLink: '#',
  helpLinkText: 'This link is for your help',
  tooltipLink: '',
  tooltipLinkText: '',
  optionalText: '',
  mainLabelTooltipBoxHeading: '',
  mainLabelTooltipBoxDescription: '',
  optionLabelTooltipBoxHeading: '',
  optionLabelTooltipBoxDescription: '',
  formik: {
    ...mockMeta,
    ...mockField,
    errors: {
      test: '',
    },
  },
};

afterEach(cleanup);
describe('[Component]: Checkbox', () => {
  //   it('should match the Snapshot', () => {
  //     const tree = renderer.create(<Checkbox {...defaultProps} />).toJSON();
  //     // console.log(render(<Checkbox {...defaultProps} />));
  //     expect(tree).toMatchSnapshot();
  //   });
  it('should render the checkbox correctly', () => {
    const props = {
      ...defaultProps,
    };
    const { getByTestId } = render(<Checkbox {...props} />);
    expect(getByTestId('checkbox-license')).toHaveTextContent('submit');
  });
});
