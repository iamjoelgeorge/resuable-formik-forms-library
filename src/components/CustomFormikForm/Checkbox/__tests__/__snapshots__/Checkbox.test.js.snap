// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Component]: Checkbox should match the Snapshot 1`] = `
<form
  action="#"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="container"
  >
    <div
      className="container  slideLabel componentHeading"
    >
      <label
        className="label"
        htmlFor="terms"
      >
        terms
      </label>
    </div>
    <div
      className="optionWithTooltip"
    >
      <label
        className="optionLabel"
      >
        <input
          checked={false}
          className="checkboxInput"
          data-testid="checkbox-terms"
          disabled={false}
          name="terms"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type="checkbox"
          value={false}
        />
        <span
          className="presentationalCheckbox"
        >
          <span
            className="checkboxChecked"
          />
        </span>
        <span
          className="labelSpanElement"
        >
          I accept
        </span>
      </label>
    </div>
    
    
    
  </div>
</form>
`;
